Source/controls/touch/renderers.cpp

```diff
@@ -0,0 +1,218 @@
+
+	int x = center.x - radius;
+	int y = center.y - radius;
+	int width = diameter;
+	int height = diameter;
+	SDL_Rect rect { x, y, width, height };
+	SDL_BlitScaled(padSurface.get(), nullptr, out.surface, &rect);
+}
+
+void VirtualDirectionPadRenderer::RenderKnob(const Surface &out)
+{
+	if (knobSurface == nullptr)
+		knobSurface.reset(LoadPNG("ui_art\\directions2.png"));
+
+	auto center = virtualDirectionPad->position;
+	auto radius = virtualDirectionPad->area.radius / 3;
+	int diameter = 2 * radius;
+
+	int x = center.x - radius;
+	int y = center.y - radius;
+	int width = diameter;
+	int height = diameter;
+	SDL_Rect rect { x, y, width, height };
+	SDL_BlitScaled(knobSurface.get(), nullptr, out.surface, &rect);
+}
+
+void VirtualPadButtonRenderer::Render(const Surface &out)
+{
+	auto center = virtualPadButton->area.position;
+	auto radius = virtualPadButton->area.radius;
+	int diameter = 2 * radius;
+
+	int x = center.x - radius;
+	int y = center.y - radius;
+	int width = diameter;
+	int height = diameter;
+
+	SDL_Surface *surface = GetButtonSurface();
+	SDL_Rect rect { x, y, width, height };
+	SDL_BlitScaled(surface, nullptr, out.surface, &rect);
+}
+
+SDL_Surface *VirtualPadButtonRenderer::GetAttackSurface()
+{
+	if (attackSurface == nullptr)
+		attackSurface.reset(LoadPNG("ui_art\\attack.png"));
+	if (pressedAttackSurface == nullptr)
+		pressedAttackSurface.reset(LoadPNG("ui_art\\attackp.png"));
+	return virtualPadButton->isHeld ? pressedAttackSurface.get() : attackSurface.get();
+}
+
+SDL_Surface *VirtualPadButtonRenderer::GetTalkSurface()
+{
+	if (talkSurface == nullptr)
+		talkSurface.reset(LoadPNG("ui_art\\talk.png"));
+	if (pressedTalkSurface == nullptr)
+		pressedTalkSurface.reset(LoadPNG("ui_art\\talkp.png"));
+	return virtualPadButton->isHeld ? pressedTalkSurface.get() : talkSurface.get();
+}
+
+SDL_Surface *VirtualPadButtonRenderer::GetItemSurface()
+{
+	if (itemSurface == nullptr)
+		itemSurface.reset(LoadPNG("ui_art\\pickitem.png"));
+	if (pressedItemSurface == nullptr)
+		pressedItemSurface.reset(LoadPNG("ui_art\\pickitemp.png"));
+	return virtualPadButton->isHeld ? pressedItemSurface.get() : itemSurface.get();
+}
+
+SDL_Surface *VirtualPadButtonRenderer::GetObjectSurface()
+{
+	if (objectSurface == nullptr)
+		objectSurface.reset(LoadPNG("ui_art\\object.png"));
+	if (pressedObjectSurface == nullptr)
+		pressedObjectSurface.reset(LoadPNG("ui_art\\objectp.png"));
+	return virtualPadButton->isHeld ? pressedObjectSurface.get() : objectSurface.get();
+}
+
+SDL_Surface *VirtualPadButtonRenderer::GetCastSurface()
+{
+	if (castSurface == nullptr)
+		castSurface.reset(LoadPNG("ui_art\\castspell.png"));
+	if (pressedCastSurface == nullptr)
+		pressedCastSurface.reset(LoadPNG("ui_art\\castspellp.png"));
+	return virtualPadButton->isHeld ? pressedCastSurface.get() : castSurface.get();
+}
+
+SDL_Surface *VirtualPadButtonRenderer::GetCancelSurface()
+{
+	if (cancelSurface == nullptr)
+		cancelSurface.reset(LoadPNG("ui_art\\back.png"));

```

Comment:
I would suggest we combine all the buttons that are drawn in the same style/size and have an enum for the various states that are represented by the contained frames (similar to other sprites/graphics/buttons).

Reply:
Can you point me to an example of one of these other sprites/graphics/buttons to give me a better idea of what you mean?

Reply:
```
enum {
     blank;
     back;
     backDown;
     pickup;
     pickupDown;
};

LoadMaskedArt("ui_art\\button.png", button, 5);

DrawArt(button, position, pressed ? pickupDown : pickup);
```